pkg/cloud/services/ec2/launchtemplate.go

```diff
@@ -530,9 +533,22 @@ func (s *Service) createLaunchTemplateData(scope scope.LaunchTemplateScope, imag
 		lt.RootVolume.DeviceName = aws.StringValue(rootDeviceName)
 
 		req := volumeToLaunchTemplateBlockDeviceMappingRequest(lt.RootVolume)
-		data.BlockDeviceMappings = []*ec2.LaunchTemplateBlockDeviceMappingRequest{
-			req,
+		blockDeviceMappings = append(blockDeviceMappings, req)
+	}
+
+	for vi := range lt.NonRootVolumes {
+		nonRootVolume := lt.NonRootVolumes[vi]
+
+		if nonRootVolume.DeviceName == "" {
+			return nil, errors.Errorf("non root volume should have device name specified")
```

Comment:
Validation should be moved to the `awsmachinepool_webhook.go`. There's already logic in other webhooks to validate volumes, you should be able to re-use.

For example: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2468/files#diff-d85ef380a29a05566d8f529b89c7746d5ce4b257400061fe52be4a9ef0647668R69-R95